Scala-style scoping traversal

samarion opened the pull request:
I based my implementation of scopes on the one in scala.tools.nsc.typechecker._ and scopes are handled automatically behind the scenes to make sure information is as faithful as possible to what's going on in the compiler.

I checked the implementation against the [MissingInterpolator pull request](https://github.com/scala/scala-abide/pull/37) and the rule can be rewritten with the code referenced from the compiler pretty much unchanged.

Old scoping rules that used scope to check tree entry paths have been moved over to PathRule subtypes since there was some clear overlap between both types of rules. Now the behavior is really different between ScopingRule and PathRule.

dragos commented:
@samarion awesome work! Thanks! I'll try to have a look at this tomorrow.

dragos commented on abide/src/main/scala/scala/tools/abide/traversal/ScopingRule.scala:
```diff
@@ -1,40 +1,1113 @@
+ *
+ * The trait provides a [[lookup]] method to find the first symbol in scope given `name` and `qualifies`
+ * filter function. Certain parts of scala scoping were ommited as not every tree is useful for scope
+ * lookup. The scoping context is complete for the following ASTs in scala.reflect.internal.Trees:
```
It would be good to mention differences from Scala scoping (if any). The list of handled ASTs looks hefty to me, so I'm wondering what's NOT handled :)

samarion replied:
Well, unfortunately the list of ASTs that are not handled is also quite hefty. Basically, the difference should be that all ASTs that are not in the list won't have associated scopes / belong to the scope.

Given the complexity of scoping in the compiler implementation, I can't even guarantee my implementation is really equivalent, but differences should be rare for the trees I listed.

dragos commented on abide/src/main/scala/scala/tools/abide/traversal/ScopingRule.scala:
```diff
@@ -1,40 +1,1113 @@
+    private[ScopingRuleTraversal] def leaveContext: State = withContext(context.parent)
+
+    /** Helper method to extract symbol from context given name and filter */
+    def lookup(name: Name, qualifies: Symbol => Boolean): NameLookup = context.lookupSymbol(name, qualifies)
```
I know it's a nitpick, but I'd like to rename `qualified` to one of `suchThat` and `p`. One is pretty standard in scalac parlance, the other one is what `filter` and others are using to designate a predicate.

samarion replied:
Okay, sounds good to me. I'll go with `suchThat` if that's what's already kind of standard in the compiler.

dragos commented:
@samarion to understand the full context (forgive the pun), what is missing in the compiler to be able to use its context implementation? If we can avoid duplicating the code, we'd be better off.

samarion commented:
That's what I was hoping to do initially, simply implement this version until I could use something from the compiler. Unfortunately, the actual scope construction in the compiler is insanely complicated because it seems to perform scoping at the same time as type inference, macro expansion, etc and even does all this in multiple passes, as far as I can tell. I'm not sure it would be feasible for the compiler to provide scoping as an API at this point. I guess we should see with @retronym and @adriaanm.

retronym commented:
I think @densh tried to reuse `Context` for experimental versions of quasiquotes. He might have even strated some refactoring work to make this possible. But it is pretty tightly coupled to typechecking so I think it proved too big a task at the time.

retronym commented:
The presentation compiler, OTOH, does perform analyses with `Context`-s. For example, code completion suggestions include the results of an implicit search. The presentation compiler has a custom `Global` that maintains a map from `Position` -> `Context` as a side-effect of every `Context` creation. Given that `abide` always runs after the typechecker in the compilation pipeline, this approach would also be feasible. We would need to refactor the context-capturing code back out of the `interactive.Global` into the regular compiler, and provide a hook by which a compiler plugin could enable this (potentially expensive) mode.

samarion commented:
Factoring out the stuff from interactive sounds promising. The extension to the nsc compiler should be pretty trivial since it already provides the `registerContext` hook. As for performance, this will anyway be much better than what I'm doing as scope is collected individually for each rule.

However, we maybe should provide the API through the internal trees since that's what we're based on. This would enable other compilers that want to support abide to still provide the necessary hooks without too much trouble as long as they offer the internal interface. WDYT?

Also, are we guaranteed after typer that each tree will point to a context, or could there be some empty positions?

densh commented:
@retronym All my experiments were quite futile due to extremely tight coupling. @vladimirnik did far more impressive job at this during his experimental attempts to move typer out of the compiler cake.

retronym commented:
@samarion The pres. compiler finds and uses the closest enclosing context for a tree. It would be a bug if one could not be found.

Statements within a block are typechecked in the same context (although if an import is introduced, a new context is spawned and used to typecheck subsequent statements.)

By way of example, not that we don't have a new context for each `Literal` tree:

```
% cat sandbox/test.scala
object Test {
  def f = {
    0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0
  }
}
% qscalac -Ydebug -Ylog:typer sandbox/test.scala 2>&1 | grep '\[context\]'
[log typer] [context] ++ NoCompilationUnit / Import(value <import>)
[log typer] [context] ++ NoCompilationUnit / Import(value <import>)
[log typer] [context] ++ NoCompilationUnit / Import(value <import>)
[log typer] [context] ++ test.scala / EmptyTree
[log typer] [context] ++ test.scala / term <empty>
[log typer] [context] ++ test.scala / term Test
[log typer] [context] ++ test.scala / term Test
[log typer] [context] ++ test.scala / term Test
[log typer] [context] ++ test.scala / Template
[log typer] [context] ++ test.scala / Template
[log typer] [context] ++ test.scala / term Test
[log typer] [context] ++ test.scala / Template
[log typer] [context] ++ test.scala / Template(value <local Test>)
[log typer] [context] ++ test.scala / term <init>
[log typer] [context] ++ test.scala / term <init>
[log typer] [context] ++ test.scala / Block
[log typer] [context] ++ test.scala / Apply
[log typer] [context] ++ test.scala / term f
[log typer] [context] ++ test.scala / Block
[log typer] [context] ++ test.scala / term f
```

retronym commented:
Actually, I think you might be able to grab the info you need by using the `pluginsTyped` extension point in the compiler plugin API.

```scala
trait AnalyzerPlugin {
    /**
     * Let analyzer plugins modify the type that has been computed for a tree.
     *
     * @param tpe   The type inferred by the type checker, initially (for first plugin) `tree.tpe`
     * @param typer The yper that type checked `tree`
     * @param tree  The type-checked tree
     * @param mode  Mode that was used for typing `tree`
     * @param pt    Expected type that was used for typing `tree`
     */
    def pluginsTyped(tpe: Type, typer: Typer, tree: Tree, mode: Mode, pt: Type): Type = tpe
```

You could record the expected type (`pt`) and context (`typer.context`) for each tree. This *might* be a bit less efficient than the IDE's approach of using positions to index into this map (with one entry per context, rather than per tree), but would be an easy place to start.

samarion commented:
This is awesome! I didn't know about that extension point.
I'll update the scoping implementation and rebase so I can test against TypeParameterShadow

retronym commented:
Be aware that trees are routinely copied, so using identity of `Tree` as the key of a `Map[Tree, Context]` might not work.

An alternatives would be to use the position, although then you might need to borrow some tricks from the presentation compiler to deal with multiple trees at the same position, and some of that code probably assumes range positions.

Or, you could attach the extra data to the `Tree` itself as a tree attachment. These are preserved through tree copying.

samarion commented:
I got it to work by using the tree attachments. I had some issues with `pluginsTyped` because it wasn't called on every tree (it's never called on TypeDefs it seems), so I used the context tree with outer to climb up the tree and assign contexts. All my tests are passing so it looks like this works.

retronym commented:
Yep, looks like `TypeDefs` that appears in type parameter lists bypass `typed1`, which is the main place that `pluginsTyped` is called. There is an extra 

```
def typedClassDef(cdef: ClassDef): Tree = {
      val clazz = cdef.symbol
      val typedMods = typedModifiers(cdef.mods)
      assert(clazz != NoSymbol, cdef)
      reenterTypeParams(cdef.tparams)
      val tparams1 = cdef.tparams mapConserve (typedTypeDef)


    def typedTypeDef(tdef: TypeDef): TypeDef =
      typerWithCondLocalContext(context.makeNewScope(tdef, tdef.symbol))(tdef.tparams.nonEmpty) {
        _.typedTypeDefImpl(tdef)
      }

    // use typedTypeDef instead. this version is called after creating a new context for the TypeDef
    private def typedTypeDefImpl(tdef: TypeDef): TypeDef = {
      tdef.symbol.initialize
      reenterTypeParams(tdef.tparams)
      val tparams1 = tdef.tparams mapConserve typedTypeDef
      val typedMods = typedModifiers(tdef.mods)
      tdef.symbol.annotations.map(_.completeInfo())

```

`ExistentialTypeTree` are treated similarly. Note the local context created in `typerWithLocalContext` to represent the scope of `forSome { ... }`.

```
      def typedExistentialTypeTree(tree: ExistentialTypeTree) = {
        val tree1 = typerWithLocalContext(context.makeNewScope(tree, context.owner)){
          typer =>
            if (context.inTypeConstructorAllowed)
              typer.context.withinTypeConstructorAllowed(typer.typedExistentialTypeTree(tree, mode))
            else
              typer.typedExistentialTypeTree(tree, mode)
        }
        checkExistentialsFeature(tree1.pos, tree1.tpe, "the existential type")
        tree1
      }
```

Your workaround looks pretty good, so long as you are carefully test that it records the innermost context for a tree. E.g, make sure that the context for `X forSome { type X }` contains `X` in scope.

samarion commented:
So, I checked for existential types which get correctly attributed with their context, but the `ExistentialTypeTree` disappear after typer and are placed in the `TypeTree.original` field. Unfortunately, it seems that they loose their tree attachments at some point, so traversal can't actually access their context. Any idea why and if this is supposed to happen?

samarion commented:
Okay, so annotating the trees didn't pan out because there were far too many holes.

I rewrote scoping to use the position-based lookup as implemented in the interactive compiler and extended the technique to work with non-range positions. I wrote some tests for both the interactive compiler trees and nsc.Global trees to make sure context lookup works everywhere. Since lookup  depends on trees having a position, some trees can't be handled so the documentation now reflects this limitation.

I also noticed quite a few rules weren't in their rule package xml definition while running tests with the sbt plugin, so I cleaned that up.

adriaanm commented:
Excellent!

SethTisue commented:
Does this PR still have a chance?
